[DO NOT MERGE][example] fold batch and sequence dimensions to accelerate Sequence Parallel

tianyu-l opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #437

Note: This PR is for showcasing purpose only and is almost a reverse of #190.

At the cost of model code change, we can obtain better Sequence Parallel performance. Without folding and unfolding,  all-gather and reduce-scatter are performed on dim 1 (sequence dim) instead of dim 0 (folded dim), which incurs an extra `aten.cat` after each collective.

Stats from @awgu:
> for 8k seq len, batch size 1 on H100, these two cats take about 0.18 ms out of 3 ms of FFN compute (6%)

Experiment on 8-layer `debug_model`
before:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/04e5ea4b-fa9e-48e5-92be-582841cb2796">
after:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/38c39506-462d-485a-a16c-48770a28edb0">

tianyu-l committed:
fold batch and sequence dimensions to accelerate Sequence Parallel

[ghstack-poisoned]

tianyu-l committed:
Update on "[DO NOT MERGE][example] fold batch and sequence dimensions to accelerate Sequence Parallel"


At the cost of model code change, we can obtain better Sequence Parallel performance. Without folding and unfolding,  all-gather and reduce-scatter are performed on dim 1 (sequence dim) instead of dim 0 (folded dim), which incurs an extra `aten.cat` after each collective.

Stats from awgu:
> for 8k seq len, batch size 1 on H100, these two cats take about 0.18 ms out of 3 ms of FFN compute (6%)

This is almost a reverse of #190.

[ghstack-poisoned]

tianyu-l committed:
Update on "[DO NOT MERGE][example] fold batch and sequence dimensions to accelerate Sequence Parallel"


Note: This PR is for showcasing purpose only and is almost a reverse of #190.

At the cost of model code change, we can obtain better Sequence Parallel performance. Without folding and unfolding,  all-gather and reduce-scatter are performed on dim 1 (sequence dim) instead of dim 0 (folded dim), which incurs an extra `aten.cat` after each collective.

Stats from awgu:
> for 8k seq len, batch size 1 on H100, these two cats take about 0.18 ms out of 3 ms of FFN compute (6%)

Experiment on 8-layer `debug_model`
before:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/04e5ea4b-fa9e-48e5-92be-582841cb2796">
after:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/38c39506-462d-485a-a16c-48770a28edb0">

[ghstack-poisoned]

awgu commented on torchtitan/parallelisms/parallelize_llama.py:
```diff
@@ -350,37 +350,41 @@ def apply_tp(model, world_mesh, parallel_dims, job_config: JobConfig):
         {
             "tok_embeddings": RowwiseParallel(
                 input_layouts=Replicate(),
-                output_layouts=Shard(1),
```
curious: could this be `output_layouts=Shard(0)` and then do not need the `PrepareModuleInput`?

tianyu-l replied:
@awgu 
Currently we are doing folding after embedding layer, so we can't do what you suggested.
But I just realize that maybe we can do folding even before embedding layer, then I think we can do this, just like the non-folding case.

tianyu-l replied:
@awgu 
OK I tried out the change. Please see comparison [here](https://github.com/pytorch/torchtitan/pull/437/files/33160f9abd47aa2133191e63e9e7849ad8a93064..346038bcd18b71befc25db4b957c1f5567f8dacb).
Everything works except the CI [failure](https://github.com/pytorch/torchtitan/actions/runs/9789630903/job/27228806203) says
> RuntimeError: It seems that we cannot capture your model as a full graph. Typical reasons include graph breaks, data/shape-dependent control flow, or missing meta kernels for custom operators. You can use our manual pipeline interfaces, or try to fix the graph breaks

So I decided to change it back.

awgu commented on torchtitan/models/llama/model.py:
```diff
@@ -187,7 +185,10 @@ def forward(
 
         """
-        bs, seqlen, _ = x.shape
+        # dim 0 of x is a folded dimension of [bs, seqlen]
```
nit: for consistency with other comments but does not matter since this is not for landing
```suggestion
        # dim 0 of x is a folded dimension of (bs, seqlen)
```

yifuwang commented:
fwiw, this can also be achieved w/ torch.compile + force_stride_order w/o changing the model code.

Basically, we can force the stride order of the all-gather/reduce-scatter input to be in a way such that `input.swapdim(0, dim)` is contiguous, and we stop enforcing the contiguity of all-gather/reduce-scatter outputs. This way, the layout transformation will be subsumed into the leading/following pointwise ops.

Async-TP currently does this ([example](https://fburl.com/5xt4vma2)). With some work we can make it work for all-gather/reduce-scatter too.

tianyu-l committed:
Update on "[DO NOT MERGE][example] fold batch and sequence dimensions to accelerate Sequence Parallel"


Note: This PR is for showcasing purpose only and is almost a reverse of #190.

At the cost of model code change, we can obtain better Sequence Parallel performance. Without folding and unfolding,  all-gather and reduce-scatter are performed on dim 1 (sequence dim) instead of dim 0 (folded dim), which incurs an extra `aten.cat` after each collective.

Stats from awgu:
> for 8k seq len, batch size 1 on H100, these two cats take about 0.18 ms out of 3 ms of FFN compute (6%)

Experiment on 8-layer `debug_model`
before:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/04e5ea4b-fa9e-48e5-92be-582841cb2796">
after:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/38c39506-462d-485a-a16c-48770a28edb0">

[ghstack-poisoned]

tianyu-l committed:
Update on "[DO NOT MERGE][example] fold batch and sequence dimensions to accelerate Sequence Parallel"


Note: This PR is for showcasing purpose only and is almost a reverse of #190.

At the cost of model code change, we can obtain better Sequence Parallel performance. Without folding and unfolding,  all-gather and reduce-scatter are performed on dim 1 (sequence dim) instead of dim 0 (folded dim), which incurs an extra `aten.cat` after each collective.

Stats from awgu:
> for 8k seq len, batch size 1 on H100, these two cats take about 0.18 ms out of 3 ms of FFN compute (6%)

Experiment on 8-layer `debug_model`
before:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/04e5ea4b-fa9e-48e5-92be-582841cb2796">
after:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/38c39506-462d-485a-a16c-48770a28edb0">

[ghstack-poisoned]

tianyu-l committed:
Update on "[DO NOT MERGE][example] fold batch and sequence dimensions to accelerate Sequence Parallel"


Note: This PR is for showcasing purpose only and is almost a reverse of #190.

At the cost of model code change, we can obtain better Sequence Parallel performance. Without folding and unfolding,  all-gather and reduce-scatter are performed on dim 1 (sequence dim) instead of dim 0 (folded dim), which incurs an extra `aten.cat` after each collective.

Stats from awgu:
> for 8k seq len, batch size 1 on H100, these two cats take about 0.18 ms out of 3 ms of FFN compute (6%)

Experiment on 8-layer `debug_model`
before:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/04e5ea4b-fa9e-48e5-92be-582841cb2796">
after:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/38c39506-462d-485a-a16c-48770a28edb0">

[ghstack-poisoned]

tianyu-l committed:
Update on "[DO NOT MERGE][example] fold batch and sequence dimensions to accelerate Sequence Parallel"


Note: This PR is for showcasing purpose only and is almost a reverse of #190.

At the cost of model code change, we can obtain better Sequence Parallel performance. Without folding and unfolding,  all-gather and reduce-scatter are performed on dim 1 (sequence dim) instead of dim 0 (folded dim), which incurs an extra `aten.cat` after each collective.

Stats from awgu:
> for 8k seq len, batch size 1 on H100, these two cats take about 0.18 ms out of 3 ms of FFN compute (6%)

Experiment on 8-layer `debug_model`
before:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/04e5ea4b-fa9e-48e5-92be-582841cb2796">
after:
<img width="1023" alt="image" src="https://github.com/pytorch/torchtitan/assets/150487191/38c39506-462d-485a-a16c-48770a28edb0">

[ghstack-poisoned]

lw commented:
Why is this marked as "example" and "do not merge"? What is the issue with this PR? Thanks!

awgu commented:
@lw because this requires changing the model code, I think @tianyu-l left it as a non-merged PR to show how people could change their own (forked) code to enable this.
